### PR TITLE
feat: show per-metric charges for paid plans

### PR DESCRIPTION
**Pull request by macko911**
## Problem

Paid plans (Starter, Growth, startup deal) show usage with no indication of what each metric costs — allowances and rates live only in Orb, and our backend has never computed or returned a dollar figure for usage.

## Solution

**Not customer-facing yet.** The Charges column is behind a dev-tool-only flag and stays off until we've verified the per-metric figures against real Orb invoices. Nothing here ships to customers on its own.

- Read Orb's per-price cost breakdown for the current billing period via `subscriptions.fetchCosts`, mapped to our metrics by billable-metric id (not price name, which is renamed per subscription)
- Add `GET /plans/billing/period-costs`, gated the same way as the upcoming-invoice endpoint
- A metric with no Orb price reads `$0.00`; an unmapped priced metric is tracked separately as `unattributedInCents` so the row shows no figure instead of a wrong one, and it's reported for visibility

Fixes [NAN-6220](https://linear.app/nango/issue/NAN-6220)

## Testing

- Unit tests for the Orb adapter's id mapping and edge cases (ended subscriptions, unattributed charges, currency mismatches), the client, and the webapp's charge-display rules
- Integration tests for the new endpoint's plan gating and Orb-response handling
- Verified end to end against live Orb (both test mode and prod-shaped fixtures) and the local dev stack: charges land on the correct row and the sum plus base fee matches the summary strip's upcoming-invoice figure

## Follow-ups

- Legacy plans, custom/enterprise pricing, and startup-deal usage credits are out of scope — deferred per the ticket

image

☝️  Note these are fake charges made on test Orb account, ignore the actual usage being 0 across all metrics.

### CODE REVIEW

**Comment by macko911** [comment]
@coderabbitai review

File: packages/billing/lib/clients/orb/adapters.ts [hunk 1]
```diff
@@ -53,6 +63,116 @@ export function fromOrbUpcomingInvoice(invoice: { amount_due: string; currency:
...
+        const priceCurrency = price.currency?.trim().toUpperCase();
+        if (!priceCurrency || !/^[A-Z]{3}$/.test(priceCurrency) || (currency && priceCurrency !== currency)) {
+            return null;
+        }
+        currency = priceCurrency;
+
+        const amountInCents = orbAmountToCents(priceCost.total);
+        if (amountInCents === null) {
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/billing/lib/clients/orb/adapters.ts]: Good catch — you're right, this silently read as "no price" and rendered $0.00 for a metric that was actually priced. Fixed by returning null for the whole conversion instead, matching the currency-mismatch handling right above it: 9bf154b8d.

File: packages/webapp/src/features/PlanOverrideOverlay.tsx [hunk 2]
```diff
@@ -188,6 +193,39 @@ export const PlanOverrideContent: React.FC<PlanOverrideContentProps> = ({ onBack
...
+                                    </SelectTrigger>
+                                    <SelectContent>
+                                        <SelectItem value={REAL_PERIOD_COSTS_VALUE}>Real charges (no override)</SelectItem>
+                                        <SelectItem value="populated">A charge on some metrics</SelectItem>
+                                        <SelectItem value="zero">$0.00 on every metric</SelectItem>
+                                        <SelectItem value="unavailable">Unavailable (no figures)</SelectItem>
+                                    </SelectContent>
+                                </Select>
+                            </div>
+                        )}
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/features/PlanOverrideOverlay.tsx:228]: This overlay is a dev-only tool for previewing simulated plan/billing states — it's gated out of production and nothing else in the file calls track() either. Adding tracking here would record synthetic dev-testing interactions as if they were real user activity, so I'm leaving it as-is.

File: packages/server/lib/services/spendAlertNotification.service.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        }
+        return Err(new Error('failed_to_send_spend_alert_emails', { cause: { accountId: team.id } }));
+    }
+
+    if (failures.length > 0) {
+        logger.warning(`Failed to send ${failures.length}/${sent.length} spend alert emails for team "${team.id}"`);
+        for (const { email, reason } of failures) {
+            report(new Error('failed_to_send_spend_alert_email', { cause: reason }), { accountId: team.id, email });
+        }
+    }
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/server/lib/services/spendAlertNotification.service.ts:116]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/webapp/src/pages/Team/Billing/components/SpendAlertDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        try {
+            await putSpendAlert({ thresholdInCents: parsed.thresholdInCents });
+            setOpen(false);
+            toast({ title: isEdit ? 'Spend alert updated' : 'Spend alert added', variant: 'success' });
+        } catch {
+            toast({ title: 'Failed to save the spend alert', variant: 'error' });
+        }
+    };
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/pages/Team/Billing/components/SpendAlertDialog.tsx:68]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/server/lib/services/spendAlertNotification.service.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        if (released.isErr()) {
+            report(released.error);
+        }
+        return Err(new Error('failed_to_send_spend_alert_emails', { cause: { accountId: team.id } }));
+    }
+
+    if (failures.length > 0) {
+        logger.warning(`Failed to send ${failures.length}/${sent.length} spend alert emails for team "${team.id}"`);
+        for (const { email, reason } of failures) {
+            report(new Error('failed_to_send_spend_alert_email', { cause: reason }), { accountId: team.id, email });
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/server/lib/services/spendAlertNotification.service.ts]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/webapp/src/pages/Team/Billing/components/SpendAlertDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    const symbol = currencySymbol(currency);
+
+    // Reset on open rather than on close, so the field starts from the saved value every time —
+    // including after that value changed while the dialog was shut.
+    useEffect(() => {
+        if (open) {
+            setAmount(isEdit ? thresholdToInput(thresholdInCents) : '');
+            setError(null);
+        }
+    }, [open, isEdit, thresholdInCents]);
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/pages/Team/Billing/components/SpendAlertDialog.tsx:52]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/server/lib/services/spendAlertNotification.service.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            if (released.isErr()) {
+                report(released.error);
+            }
+            return Err(new Error('failed_to_resolve_spend_alert_recipients', { cause: { accountId: team.id } }));
+        }
+
+        // Marked rather than released: there is nobody to tell, and that won't change within the
+        // period, so a redelivery should not redo this lookup to reach the same conclusion.
+        logger.warning(`No spend alert recipients for team "${team.id}"`);
+        const marked = await markSpendAlertNotified(db.knex, claim);
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/server/lib/services/spendAlertNotification.service.ts:82]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/server/lib/services/spendAlertNotification.service.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        if (marked.isErr()) {
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/server/lib/services/spendAlertNotification.service.ts:83]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/server/lib/services/spendAlertNotification.service.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    if (customer.isOk()) {
+        emails.add(customer.value.invoicingDetails.email.toLowerCase());
+        for (const additional of customer.value.invoicingDetails.additionalEmails) {
+            emails.add(additional.toLowerCase());
+        }
+    } else {
+        report(customer.error, { accountId: team.id });
+    }
+
+    for (const admin of await userService.getVerifiedActiveAdministratorsByAccountId(team.id)) {
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/server/lib/services/spendAlertNotification.service.ts:150]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/webapp/src/pages/Team/Billing/components/SpendAlerts.tsx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        void confirm({
+            title: 'Remove spend alert',
+            description: "You'll stop being emailed when spend crosses this amount.",
+            confirmButtonText: 'Remove',
+            confirmVariant: 'danger',
+            onConfirm: async () => {
+                try {
+                    await deleteSpendAlert();
+                    toast({ title: 'Spend alert removed', variant: 'success' });
+                } catch {
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/pages/Team/Billing/components/SpendAlerts.tsx:34]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/database/lib/migrations/20260821090000_create_spend_alert_notifications.cjs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+exports.config = { transaction: false };
+
+/**
+ * @param {import('knex').Knex} knex
+ */
+exports.up = async function (knex) {
+    await knex.raw(`
+        CREATE TABLE IF NOT EXISTS spend_alert_notifications (
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/database/lib/migrations/20260821090000_create_spend_alert_notifications.cjs:8]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/shared/lib/services/plans/spendAlertNotifications.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+export const SPEND_ALERT_NOTIFICATIONS_TABLE = 'spend_alert_notifications';
+
+/**
+ * How long a claim stays another caller's before it can be taken over.
+ *
+ * Matches Orb's own webhook timestamp tolerance. Shorter and a slow send risks being notified
+ * twice; longer and a crashed one stays silent for that much of the billing period.
+ */
+export const SPEND_ALERT_CLAIM_LEASE_MINUTES = 5;
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/shared/lib/services/plans/spendAlertNotifications.ts:11]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

File: packages/webapp/src/pages/Team/Billing/components/SpendAlerts.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
+import { Pencil, Plus, Trash2 } from 'lucide-react';
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/pages/Team/Billing/components/SpendAlerts.tsx:1]: This PR was briefly opened against `master` before I noticed NAN-6222's spend-alert work is still an open PR (#7180) and retargeted onto it. This comment landed on a file that isn't part of this PR's actual diff, so it doesn't apply here — if it's still relevant, it belongs on #7180.

**Comment by macko911** [comment]
@coderabbitai review

File: packages/webapp/src/pages/Team/Billing/useSelectedMonth.ts [hunk 14]
```diff
@@ -54,5 +56,5 @@ export function useSelectedMonth(): UseSelectedMonth {
 
     const canGoPrevious = useMemo(() => selectedMonth.getTime() > EARLIEST_USAGE_MONTH_MS, [selectedMonth]);
 
-    return { selectedMonth, setSelectedMonth, canGoNext, canGoPrevious };
+    return { selectedMonth, setSelectedMonth, canGoNext, canGoPrevious, isCurrentMonth: !canGoNext };
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/pages/Team/Billing/useSelectedMonth.ts:58]: Real gap, but pre-existing and out of scope here: this same staleness pattern already exists in FreeUsage.tsx today (isCurrentMonth memoized only on selectedMonth, unrelated to this PR) — I only lifted it into this shared hook. It's a narrow window (needs the tab to stay open across a UTC month boundary with zero navigation) and self-corrects on any interaction. Not fixing it as part of a charges ticket.

File: packages/webapp/src/pages/Team/Billing/components/Usage.tsx [hunk 15]
```diff
@@ -39,6 +41,13 @@ export const Usage: React.FC = () => {
     // billing running-average, matching what each row's drill-in chart also requests.
     const { data: usage, isLoading, error: usageError } = useApiGetBillingUsage(env, timeframe, { avgPerDay: true, enabled: plan != null && !isFree });
 
+    // Behind a dev-tool flag until the figures are checked against real Orb invoices.
+    const metricChargesEnabled = usePlanOverrideStore((s) => s.metricChargesEnabled);
+    // Orb only holds costs for the period in progress, so a past month has no charge to state.
+    const chargesEnabled = metricChargesEnabled && isCurrentMonth && hasMonthlySpend(plan);
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/pages/Team/Billing/components/Usage.tsx:46]: Real gap, but pre-existing and out of scope here: this same staleness pattern already exists in FreeUsage.tsx today (isCurrentMonth memoized only on selectedMonth, unrelated to this PR) — I only lifted it into this shared hook. It's a narrow window (needs the tab to stay open across a UTC month boundary with zero navigation) and self-corrects on any interaction. Not fixing it as part of a charges ticket.

**Comment by macko911** [comment]
@coderabbitai review

File: packages/billing/lib/clients/orb/adapters.ts [hunk 16]
```diff
@@ -43,25 +44,110 @@ export function orbAmountToCents(amount: string): number | null {
...
+/** Fixed prices are excluded, so the metrics never sum to the period's invoice total. */
+export function fromOrbPeriodCosts(costs: { data: OrbCostBucket[] }, now: Date): BillingPeriodCosts | null {
+    // Cumulative buckets accumulate over the period, so the one ending last spans all of it.
+    const period = costs.data.reduce<OrbCostBucket | null>(
+        (latest, bucket) => (latest && Date.parse(latest.timeframe_end) >= Date.parse(bucket.timeframe_end) ? latest : bucket),
+        null
+    );
+    // An ended subscription still answers with its final period rather than erroring, and those costs
+    // are not what is being billed now.
+    if (!period || Date.parse(period.timeframe_end) <= now.getTime()) {
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/billing/lib/clients/orb/adapters.ts]: Good catch — fixed. `NaN <= now.getTime()` is always false, so a malformed `timeframe_end` would have slipped through as a current period instead of being rejected. Added an explicit `Number.isNaN` check plus a test: 2d2210f92.

File: packages/webapp/src/pages/Team/Billing/components/Usage.tsx [hunk 17]
```diff
@@ -39,6 +41,12 @@ export const Usage: React.FC = () => {
     // billing running-average, matching what each row's drill-in chart also requests.
     const { data: usage, isLoading, error: usageError } = useApiGetBillingUsage(env, timeframe, { avgPerDay: true, enabled: plan != null && !isFree });
 
+    const metricChargesEnabled = usePlanOverrideStore((s) => s.metricChargesEnabled);
+    // Orb only holds costs for the period in progress, so a past month has no charge to state.
+    const chargesEnabled = metricChargesEnabled && isCurrentMonth && hasMonthlySpend(plan);
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/pages/Team/Billing/components/Usage.tsx:46]: This one's the same UTC-month-rollover concern raised earlier in this PR (both you and CodeRabbit flagged it before the last rebase). We looked at it then: it's a real but pre-existing gap — `FreeUsage.tsx` has the identical staleness pattern already, unrelated to this PR, and the window only opens if the tab stays untouched across a UTC month boundary, self-correcting on any navigation or refocus. Leaving it out of scope here, same call as before.

File: packages/billing/lib/clients/orb/client.ts [hunk 18]
```diff
@@ -218,6 +227,30 @@ export class OrbClient implements BillingClient {
...
+                    headers: {
+                        'Orb-Cache-Control': 'cache',
+                        'Orb-Cache-Max-Age-Seconds': '300'
+                    }
+                }
+            );
+
+            return Ok(fromOrbPeriodCosts(costs, new Date()));
+        } catch (err) {
+            if (isOrbNotFoundError(err)) {
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/billing/lib/clients/orb/client.ts:259]: Checked this against real Orb behavior rather than guessing: I hit `fetchCosts` directly against two subscriptions confirmed 'status: ended' in Orb, and both returned 200 with their final period's data — not a 400. So `fetchCosts` doesn't seem to throw the 'status ended' error `fetchUpcoming` does; `isOrbEndedSubscriptionError`'s string match is also specific to `fetchUpcoming`'s wording, so it likely wouldn't match a `fetchCosts` error even if one existed. The ended-subscription case is already handled — `fromOrbPeriodCosts`'s own `timeframe_end <= now` guard treats that final-period response as closed and returns null (covered by a test: 'returns null for a period that has already closed'). Leaving this as-is rather than adding a check for an error path I can't confirm ever fires.

File: packages/webapp/src/features/planOverride.ts [hunk 19]
```diff
@@ -86,6 +103,32 @@ export function buildSpendOverride(override: SpendOverride): GetUpcomingInvoice[
     return { data: { amountInCents: override, currency: 'USD' } };
 }
 
+/**
+ * `populated` omits `records` entirely: a metric with no price at all is a state real accounts are
+ * in, and it is not the same as a zero charge.
+ */
```
[the remark this answers is not in the document]
Comment (macko911) [reply packages/webapp/src/features/planOverride.ts:109]: I think this one's a misread — `records` is a real `UsageMetric` (it's one of the 8 literal values in that union), it's just the one deliberately left out of this fixture's `metrics` object to exercise the 'no price for this metric' display case. The comment's claim holds: `populated` really does omit it while populating the other six. Leaving the comment as-is.

**Comment by macko911** [comment]
@coderabbitai review

File: packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts [hunk 20]
```diff
@@ -0,0 +1,62 @@
+import { billing } from '@nangohq/billing';
+import { metrics, report, requireEmptyQuery, zodErrorToHTTP } from '@nangohq/utils';
+
+import { asyncWrapper } from '../../../../utils/asyncWrapper.js';
+import { isSpendPlan } from '../../../../utils/spendPlans.js';
+
+import type { GetBillingPeriodCosts } from '@nangohq/types';
+
+const NO_COSTS = { metrics: {}, unattributedInCents: null, currency: null };
```
Comment (pfreixes) [inline packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts]: do we like returning no costs associated as an empty object? I know that we have full control of the request flow .... but wondering if it can lead to subtle bugs where we coud be returning an empty object and believing that is a NO_COSTS one 

Wondering if would be worth it to have a field in our response for this

File: packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+    const costsRes = await billing.getPeriodCosts(plan.orb_subscription_id);
+    if (costsRes.isErr()) {
+        report(costsRes.error);
+        res.status(500).send({ error: { code: 'server_error', message: 'Failed to get period costs' } });
+        return;
+    }
+
+    const costs = costsRes.value;
+    if (!costs) {
+        res.status(200).send({ data: NO_COSTS });
```
Comment (pfreixes) [inline packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts:45]: should this ever happen?

File: packages/billing/lib/clients/orb/adapters.ts [hunk 22]
```diff
@@ -43,25 +44,112 @@ export function orbAmountToCents(amount: string): number | null {
...
+            continue;
+        }
+
+        const priceCurrency = normalizeIsoCurrency(price.currency);
+        if (!priceCurrency || (currency && priceCurrency !== currency)) {
+            return null;
+        }
+        currency = priceCurrency;
+
+        // One unparseable amount makes the whole response untrustworthy: the metrics that did parse
```
Comment (pfreixes) [inline packages/billing/lib/clients/orb/adapters.ts]: should this just throw and return a 500? so why the response should be invalid? just trying to understand.

If orb can not provide a right response, shouldn't this be a transient error from their side, and we fail fast? if we return null we just return an empty object to our dashboard, is it the behavioour that we want?

File: packages/billing/lib/clients/orb/adapters.ts [hunk 23]
```diff
@@ -43,25 +44,112 @@ export function orbAmountToCents(amount: string): number | null {
...
+        // would understate the period.
+        const amountInCents = orbAmountToCents(priceCost.total);
+        if (amountInCents === null) {
+            return null;
+        }
+
+        const metric = price.billable_metric ? orbBillableMetricToUsageMetric[price.billable_metric.id] : undefined;
+        if (!metric) {
```
Comment (pfreixes) [inline packages/billing/lib/clients/orb/adapters.ts:147]: should we return the metrics that we could not attribute? so we log them or we make them part of the metric? otherwise if we have two we will not know this.

Also should we keep the `unattributedInCents` just contained to this method rather than having the handler to deal with this?

File: packages/billing/lib/clients/orb/adapters.ts [hunk 24]
```diff
@@ -43,25 +44,112 @@ export function orbAmountToCents(amount: string): number | null {
...
     }
 
     return { amountInCents, currency };
 }
 
+/**
+ * Prod and test mode share no billable-metric ids, so both sets live here. Names can't be the key:
+ * one metric bills under several of them.
+ */
+const orbBillableMetricToUsageMetric: Record<string, UsageMetric> = {
```
Comment (pfreixes) [inline packages/billing/lib/clients/orb/adapters.ts:72]: Food for thought, is there a way to avoid having this dictionary, so we do not need to update it every time that we add a new billable metric? Wondering if we could use the `extenal_price_id` or the metadata for enriching our prices or billable metrics in Orb with the value that would allow us to just use the `UsageMetric`

Links:

https://docs.withorb.com/api-reference/price/list-prices
https://docs.withorb.com/api-reference/metric/update-metric
https://docs.withorb.com/api-reference/price/update-price

As said just food for thought

**Review by pfreixes — COMMENTED** [review]
Left some comments! let me know how I can help on speeding up the release of this one

File: packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts (hunk 20 above)
Comment (macko911) [reply packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts]: Added an explicit `noCosts: boolean` to the response instead of inferring it from null fields.

Fixed in https://github.com/NangoHQ/nango/commit/7485043d8

File: packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts (hunk 21 above)
Comment (macko911) [reply packages/server/lib/controllers/v1/plans/billing/getBillingPeriodCosts.ts:45]: Yes — see the null returns here: https://github.com/NangoHQ/nango/blob/7485043d8/packages/billing/lib/clients/orb/adapters.ts#L100-L164. Mainly an ended subscription (2 such accounts in the prod sweep) or an empty/closed period.

File: packages/billing/lib/clients/orb/adapters.ts (hunk 22 above)
Comment (macko911) [reply packages/billing/lib/clients/orb/adapters.ts]: Landed on neither, in the end. Not a 500 — a bad price on this subscription isn't a transient failure we'd recover by retrying, so failing the whole request just to signal it felt disproportionate.

But also not "the whole response goes blank" — that was the part I got wrong initially. A price's `billable_metric` id tells us which row it belongs to independently of whether its amount or currency parsed, so one bad price now only dashes its own metric. Every other metric in the same response keeps its real number. If a price is bad *and* we can't tell which metric it's for, that's the one case where we still can't rule out any unpriced row — same as an unmapped price always has.

Reported either way, so we find out without the customer's dashboard breaking over it.

Fixed in https://github.com/NangoHQ/nango/commit/7485043d8

File: packages/billing/lib/clients/orb/adapters.ts (hunk 23 above)
Comment (macko911) [reply packages/billing/lib/clients/orb/adapters.ts:147]: Both done. The report now carries the price id, name, and metric (when known) for every price that didn't make it into a clean number, not just a summed total — so two bad prices are distinguishable.

And the reporting moved out of the HTTP handler into `OrbClient.getPeriodCosts`, next to the client's other Orb telemetry — it already does its own `metrics.increment` for event ingestion, so this follows that precedent.

Fixed in https://github.com/NangoHQ/nango/commit/7485043d8

File: packages/billing/lib/clients/orb/adapters.ts (hunk 24 above)
Comment (macko911) [reply packages/billing/lib/clients/orb/adapters.ts:72]: Good idea, I agree the hard-coded IDs are not an elegant solution but I'd ship it in the first iteration if you don't object.
I plan to keep this feature behind a flag for the time being and we will need to update the id list once we move onto the new pricing next month.
Filed [NAN-6755](https://linear.app/nango/issue/NAN-6755/move-the-orb-billable-metric-mapping-out-of-code) to track it.

**Review by pfreixes — APPROVED** [review]
